### ISSUE TITLE
ssh session open error: 'client not avilable' using salt provisioner

### ISSUE THREAD

**Issue by vectorsigma** [hashicorp/terraform#17413]
I'm trying to use the salt-masterless provisioner to provision a single EC2 instance according to the states repo configured.

I'm able to ssh into the spawned instance, but the terraform binary doesn't appear to be able to, from the same machine.  I'm using the latest release of terraform, as it fixed the only issue that seemed related.

### Terraform Version

```
$ terraform -version
Terraform v0.11.3                                                                                                                                                                                    
+ provider.aws v1.9.0
```

### Terraform Configuration Files
```
resource "aws_instance" "cal1" {
  ami = "${data.aws_ami.f27.id}"
  instance_type = "${var.instance_type}"
  key_name = "${var.ssh_key}"
  availability_zone = "${var.aws_region}a"
  vpc_security_group_ids = [ "${aws_security_group.only_me.id}" ]
  subnet_id = "${aws_subnet.base_vpc_az1.id}"

  provisioner "salt-masterless" {
    local_state_tree = "${var.salt_path}/salt"
    local_pillar_roots = "${var.salt_path}/pillar"
  }

  tags {
    Name = "calendar-server"
    SaltRole = "calendar-server"
  }
}
...
```

### Debug Output

[debug output](https://gist.github.com/vectorsigma/b8a7165fbd05a262cfe731ab598f74f6)

### Expected Behavior

EC2 instance would be provisioned using the saltmaster.

### Actual Behavior

Exits with an error: 
```
Error: Error applying plan:

1 error(s) occurred:

* aws_instance.cal1: 1 error(s) occurred:

* Unable to install Salt: dial tcp 34.213.194.235:22: i/o timeout

```

### Steps to Reproduce
1. `terraform plan -var-file=terraform.tfvars apply.out`
2. `terraform apply apply.out`

### Additional Context

I have set my AWS_PROFILE environment variable.  I'm also executing the terraform commands via a make file, but the target basically just runs: terraform init, terraform plan, then terraform apply. Not complicated.

### References

- #16678

**Comment by vectorsigma** [comment]
FWIW, the salt-masterless provisioner in Packer 1.2.0 works perfectly.  I don't know off hand if either project shares any code, but that might be worth investigating.

**Comment by jbardin** [comment]
Hi @vectorsigma,

Thanks for filing the issue. The salt-masterless provisioner did not retry to connection if the host was not yet responsive.  This should be fixed in master by #17434.
